### DISCUSSION TITLE
How to deploy cdk8s applications

### DISCUSSION THREAD

**Discussion by iliapolo** [cdk8s-team/cdk8s#1199] [Ideas]
From its inception, cdk8s was designed to be a Kubernetes manifest authoring tool. Since cdk8s generates pure Kubernetes manifests, deployment of cdk8s applications can be achieved by passing the output of cdk8s synth to one the great and popular deployment tools that already exist in the ecosystem.

In some cases, integrating with such deployment tools is fairly straight forward. For example, one can pass the dist directory created by cdk8s directly to kubectl and deploy the set of manifests: `kubectl apply -f ./dist`. In other cases, it can be more complex. For example, deploying with Helm requires more work because Helm expects a Helm chart as its input.

While we still believe this is the right approach, we would like to make it easier for customers to integrate cdk8s with existing deployment tools and pipelines. We think we can achieve that by providing guides and other written content which we we feel is currently missing. The following is a list of articles we would like to publish:

* Deploy cdk8s applications with Helm.
* Deploy cdk8s applications with Kubectl.
* Build a GitOps pipeline with Flux and cdk8s.
* Build a GitOps pipeline with ArgoCD and cdk8s.
* Multi-environment manifests with cdk8s.

We would love to hear your thoughts on what types of integrations or content you would like to see in cdk8s? What would make deploying cdk8s application easier?
In addition, we want to publish a cdk8s best practices guide that will talk about:

* Should manifests be committed to source?
* How to write unit and integration tests?
* How to structure the app?
* How to run policy-as-code manifest validations?
* How to create organizational defaults and abstractions?

We would love to hear if there any other questions related to maintaining a cdk8s application you would like us to address.

**Comment by rothgar** [comment]
Is there an opportunity or a desire to make self deploying apps? eg I have a node app, I can import cdk8s library and never worry about k8s manifests or installing kubectl.

I think having the content outlined will be great, but I also feel like there's strength in cdk8s that's not being utilized

**Reply by iliapolo** [reply]
Can you elaborate what you mean by "self-deploying"? Are you referring to a possible `cdk8s deploy` command?

**Reply by rothgar** [reply]
More like a way to import cdk8s library and not have to use the cdk8s (or kubectl) cli. Here's an old, but good demo of what I was thinking of https://www.youtube.com/watch?v=XPC-hFL-4lU

**Reply by shinebayar-g** [reply]
I think you can use `app.synthYaml()` method to grab the yaml output and use it in your nodejs app. You lose some context information though, like which chart it's coming from etc.. But deployment is not a responsibility of the cdk8s, it's up to you to implement.

**Comment by shinebayar-g** [comment]
- Build a GitOps pipeline with ArgoCD and cdk8s.
- Multi-environment manifests with cdk8s.

I think this would be 🔥 Currently trying to navigate around how to manage multi environment, multi cluster apps in a mono repo. Things are little bit easier to maintain for us in a monorepo at this moment. But would like to see how it scales in a large scale deployment.

**Comment by louisdussarps** [comment]
A bit late to the party, but for those who would like to manage their deployments programmatically, we have published a deployment tool that brings a CloudFormation-like experience to CDK8s through TypeScript workflows. It handles rollbacks on failed deployments and supports exporting values from charts.
Here is the github repository : https://github.com/LaWebcapsule/orbits
We also wrote a blog post on this topic : https://orbits.do/blog/programmable-cdk8s-deployment
